docs for branded types (#112)
pelotom/runtypes: README.md

divide(10, 0); // Throws error: division by zero
```

## Branded types

Branded types is a way to emphasize the uniqueness of a type:

```ts
const Login = String.withBrand(
  'login',
);
const Password = String.withBrand(
  'password',
  // optional constraint
  str => str.length >= 8 || 'Too short password',
);

const fetchSignIn = Contract(
  Login,
  Password,
  Unknown,
).enforce((login, pass) => {/*...*/});

const login = Login.check('myLogin');
const password = Password.check('12345678');

// static type ok, runtime ok
fetchSignIn(login, password);

// static type error, runtime error
fetchSignIn(password, login);

// static type error, runtime ok
fetchSignIn('a@b.c', '12345678');
```

> Branded types is like [opaque types](https://flow.org/en/docs/types/opaque-types) and work as expected, except it impossible to use branded strings as a key, [before nominal types release](https://github.com/microsoft/TypeScript/pull/33038), because currently, a brand is setting by tricky workaround by adding unique symbol to the main type:

Reviewer: ```suggestion
Branded types are like [opaque types](https://flow.org/en/docs/types/opaque-types) and work as expected, except it is impossible to use branded strings as a key [until we have nominal types](https://github.com/microsoft/TypeScript/pull/33038). Currently the workaround is to add a unique symbol to the main type:
```

```ts

> const StringBranded = String.withBrand(
>   'my brand',
> )
> type StringBrandedStatic =  Static<typeof StringBranded>
> 
> type Prop = { [K in StringBrandedStatic]: 'prop' }[
>   // `Type 'string & { [RuntypeName]: "my brand"; }' cannot be used as an index type.`
>   StringBrandedStatic
> ]

Reviewer 2: ```suggestion
const StringBranded = String.withBrand('StringBranded');
type StringBranded = Static<typeof StringBranded>;
// Then the type `StringBranded` is computed as:
// string & { [RuntypeName]: "StringBranded" }
```
